File: lib/py/hpc.py
#!/usr/bin/env python3
"""Shared HPC helpers: cluster/node resolution + Kerberos auth.

Imported by both cli.py (cp) and sshfs.py. Kept in its own module so they can
share without a circular import — cli.py runs as __main__, so `from cli import …`
would not resolve. Config comes from the inherited (exported) shell env.
"""
import os
import subprocess
from typing import List

import typer

from log import mu_log


def cluster_defs():
    """Yield (cluster, domain, [nodes]) from the inherited env config."""
    for c in os.environ.get("MU_CLUSTERS", "").split():
        cu = c.upper()
        domain = os.environ.get(f"MU_CLUSTER_{cu}_DOMAIN", "")
        if not domain:
            continue
        yield c, domain, os.environ.get(f"MU_CLUSTER_{cu}_NODES", "").split()


def node_targets() -> dict:
    """Map every node name -> user@node.domain."""
    user = os.environ.get("MU_HPC_UNAME", "")
    return {
        node: f"{user}@{node}.{domain}"
        for _, domain, nodes in cluster_defs()
        for node in nodes
    }


def complete_node(incomplete: str) -> List[str]:
    return [n for n in node_targets() if n.startswith(incomplete)]


def nodes_epilog() -> str:
    """Live node list for --help, read from the inherited env at import time."""
    nodes = sorted(node_targets())
    if not nodes:
        return "No nodes configured (is MU_CLUSTERS set?)."
    return "Known nodes: " + ", ".join(nodes) + "  ·  see 'mu cp nodes' for targets."


def resolve(node_or_target: str) -> str:
    """Accept a bare node name (`mike`) or an explicit `user@host` target."""
    if "@" in node_or_target:
        return node_or_target
    targets = node_targets()
    if node_or_target not in targets:
        known = ", ".join(sorted(targets)) or "(none — is MU_CLUSTERS set?)"
        mu_log("ERROR", f"unknown node: {node_or_target} (known: {known})")
        raise typer.Exit(2)
    return targets[node_or_target]


def ensure_ticket() -> None:
    """Obtain a Kerberos ticket if none is present. In-body so --help/completion
    never trigger pkinit."""
    user = os.environ.get("MU_HPC_UNAME", "")
    if not user:
        return
    try:
        klist = subprocess.run(["klist"], capture_output=True, text=True)
    except FileNotFoundError:
        return  # no kerberos here (e.g. hpc login node) — nothing to do
    if user in klist.stdout:
        return
    mu_log("INFO", f"No Kerberos ticket for {user}; running pkinit…")
    subprocess.run(["pkinit", user])

File: internal/hpc/hpc.go
// Package hpc wraps the cluster inventory (internal/config) with the behaviors
// the transfer plane needs: node resolution, tab-completion, and Kerberos ticket
// acquisition. Mirrors the retired Python hpc.py; kept apart from config so the
// low-level env reader stays free of subprocess/auth side effects.
package hpc

import (
	"fmt"
	"os"
	"os/exec"
	"sort"
	"strings"
	"time"

	"github.com/mayhl/mayhl_utils/internal/config"
	"github.com/mayhl/mayhl_utils/internal/render"
)

// Resolve accepts a bare node name (e.g. "hpc2") or an explicit "user@host"
// target. A bare name must be a configured node; otherwise an error listing the
// known nodes is returned.
func Resolve(nodeOrTarget string) (string, error) {
	if strings.Contains(nodeOrTarget, "@") {
		return nodeOrTarget, nil
	}
	if t, ok := config.NodeTargets()[nodeOrTarget]; ok {
		return t, nil
	}
	known := strings.Join(config.NodeNames(), ", ")
	if known == "" {
		known = "(none — is MU_CLUSTERS set?)"
	}
	return "", fmt.Errorf("unknown node: %s (known: %s)", nodeOrTarget, known)
}

// CompleteNode returns configured node names that start with prefix, for shell
// completion of the node argument.
func CompleteNode(prefix string) []string {
	var out []string
	for n := range config.NodeTargets() {
		if strings.HasPrefix(n, prefix) {
			out = append(out, n)
		}
	}
	sort.Strings(out)
	return out
}

// NodesHint is a one-line "Known nodes: …" string for command help, read from
// the inherited env at construction time.
func NodesHint() string {
	names := config.NodeNames()
	if len(names) == 0 {
		return "No nodes configured (is MU_CLUSTERS set?)."
	}
	return "Known nodes: " + strings.Join(names, ", ") + "  ·  see 'mu hpc nodes' for targets."
}

// EnsureTicket obtains a Kerberos ticket if the cache lacks a usable one —
// present, matching, and not expired or about to (an expired ticket still lists
// its principal, so mere presence isn't enough). Called in the command body (not
// at construction) so --help and completion never trigger pkinit. It's a no-op on
// an HPC login/compute node ($BC_HOST set): the ticket is already there from
// login, so we never touch Kerberos there — mirroring the shell auth seam
// (mu_auth is `:` on hpc, pkinit on local). Also a no-op when MU_HPC_UNAME is
// unset or klist is absent. A failed pkinit (offline, CAC/PIN) is an error the
// caller must abort on — proceeding just buries the cause under an ssh timeout.
func EnsureTicket() error {
	if os.Getenv("BC_HOST") != "" || os.Getenv("MU_SYSTEM") == "hpc" {
		return nil
	}
	user := config.HPCUser()
	if user == "" {
		return nil
	}
	info, hasKlist := Ticket()
	if !hasKlist || ticketUsable(info, user, time.Now()) {
		return nil
	}
	render.Info(fmt.Sprintf("No valid Kerberos ticket for %s; running pkinit…", user))
	cmd := exec.Command("pkinit", user)
	cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("pkinit failed — no valid Kerberos ticket for %s (offline, or a CAC/PIN problem?)", user)
	}
	return nil
}
